opencensus_shim/src/main/java/io/opentelemetry/opencensusshim/OpenTelemetryTraceComponentImpl.java

```diff
@@ -0,0 +1,62 @@
+/*
+ * Copyright The OpenTelemetry Authors
+ * SPDX-License-Identifier: Apache-2.0
+ */
+
+package io.opentelemetry.opencensusshim;
+
+import io.opencensus.common.Clock;
+import io.opencensus.implcore.common.MillisClock;
+import io.opencensus.implcore.trace.RecordEventsSpanImpl.StartEndHandler;
+import io.opencensus.implcore.trace.TracerImpl;
+import io.opencensus.implcore.trace.config.TraceConfigImpl;
+import io.opencensus.implcore.trace.internal.RandomHandler;
+import io.opencensus.implcore.trace.propagation.PropagationComponentImpl;
+import io.opencensus.trace.TraceComponent;
+import io.opencensus.trace.Tracer;
+import io.opencensus.trace.config.TraceConfig;
+import io.opencensus.trace.export.ExportComponent;
+import io.opencensus.trace.propagation.PropagationComponent;
+
+/** Java 8 implementation of the {@link TraceComponent} for OpenTelemetry migration. */
+public final class OpenTelemetryTraceComponentImpl extends TraceComponent {
+  private final PropagationComponent propagationComponent = new PropagationComponentImpl();
+  private final ExportComponent noopExportComponent = ExportComponent.newNoopExportComponent();
+  private final Clock clock;
+  private final TraceConfig traceConfig = new TraceConfigImpl();
+  private final Tracer tracer;
+
+  /** Public constructor to be used with reflection loading. */
+  public OpenTelemetryTraceComponentImpl() {
+    clock = MillisClock.getInstance();
+    RandomHandler randomHandler = new ThreadLocalRandomHandler();
+    StartEndHandler startEndHandler = new OpenTelemetryStartEndHandler();
+    tracer = new TracerImpl(randomHandler, startEndHandler, clock, traceConfig);
```

Comment:
Instead of using opencensus `TracerImpl`, can you use a shim implementation that delegates to a opentelemetry tracer? This will remove the need for the `SpanCache` which looks like it can affect performance a lot, the custom OC Tracer would return OC Spans that just wrap the OTel spans. FWIU, the opentracing shim takes such an approach so might be a good reference.

Reply:
Yeah I [tried a solution](https://github.com/census-instrumentation/opencensus-java/compare/master...zoercai:migration-create-new-span-in-otel-context) where I changed OpenCensus so that the shim would turn on another path in span creation, where an OTel tracer would take over to create the span and handle the rest. We didn't end up going with this approach as we didn't want OpenCensus to depend on OpenTelemetry, especially since we're not really maintaining OpenCensus anymore and don't want to do any more releases on OpenCensus unless there is some critical bug.

Reply:
I haven't looked into opentracing much, but my understanding is that the opentracing shim is able to do so as opentracing is mainly interfaces that can be overridden, whereas with OpenCensus we'd actually have to change the implementation inside as well

Reply:
Apologies if this is basic since I didn't dig into enough but from my understanding here we have `Tracer` which is an interface but we are returning `TracerImpl` from OC. Doesn't returning an implementation backed by OTel reduce the OC in the app instead of increasing it? 

IIUC the goal of the project is to allow OC users to gracefully migrate to OTel. So I guess the first step is to implement the OC API using OTel - then OC API usage can be migrated to OTel API usage gradually with no practical effect since the implementation is still OTel, the OC API dependency is finally removed and user has fully migrated. If this isn't the case, I think I'm missing a big point so let me know :)

Reply:
1. Tracer is an [abstract class not an interface](https://github.com/census-instrumentation/opencensus-java/blob/660e8f375bb483a3eb817940b3aa8534f86da314/api/src/main/java/io/opencensus/trace/Tracer.java#L72)
2. Span is an [abstract class not an interface](https://github.com/census-instrumentation/opencensus-java/blob/660e8f375bb483a3eb817940b3aa8534f86da314/api/src/main/java/io/opencensus/trace/Span.java#L37)
3. In addition, I believe there are some static methods which may rely on certain implementation details which can't be easily swapped out with custom code using OT.

Given that OpenCensus was never designed to support custom SDK implementation (unlike OpenTracing and OpenTelemetry), it makes it pretty hard to completely replace the implementation.

So what we did instead is replaced the low-level `Context` management to use OT under the hood and also introduced a `StartEndHandler` hook which allows us to mirror all changes on OC side to OT. The rest of the code behaves as if there is a standard OC implementation.

Reply:
Thanks for the context, makes sense especially given the Span not being an interface